src/trs/lexicon.rs

```diff
@@ -1,162 +1,182 @@
-///! Hindley-Milner Typing for First-Order Term Rewriting Systems (no abstraction)
-///!
-///! Much thanks to:
-///! - https://github.com/rob-smallshire/hindley-milner-python
-///! - https://en.wikipedia.org/wiki/Hindley%E2%80%93Milner_type_system
-///! - (TAPL; Pierce, 2002, ch. 22)
 use itertools::Itertools;
-use polytype::{self, Context, Type, TypeSchema, Variable as TypeVar};
+use polytype::{Context, Type, TypeSchema, Variable as TypeVar};
 use rand::{thread_rng, Rng};
 use std::f64::NEG_INFINITY;
 use std::fmt;
 use std::iter::{once, repeat};
-use term_rewriting::{Atom, MergeStrategy, Operator, Rule, Signature, Term, Variable, TRS as UntypedTRS};
+use std::sync::{Arc, RwLock};
+use term_rewriting::{Atom, Operator, Rule, Signature, Term, Variable, TRS};
 
-use super::trace::Trace;
+use super::{RewriteSystem, SampleError, TypeError};
 use utils::logsumexp;
 use GP;
 
 #[derive(Debug, Clone)]
-pub enum TypeError {
-    Unification(polytype::UnificationError),
-    OpNotFound,
-    VarNotFound,
+pub struct GeneticParams {
+    pub h0: RewriteSystem,
+    pub n_crosses: usize,
 }
-impl From<polytype::UnificationError> for TypeError {
-    fn from(e: polytype::UnificationError) -> TypeError {
-        TypeError::Unification(e)
+
+/// Manages the syntax of a term rewriting system.
+#[derive(Clone)]
+pub struct Lexicon(pub(crate) Arc<RwLock<Lex>>);
+impl Lexicon {
+    /// All the free type variables in the lexicon.
+    pub fn free_vars(&self) -> Vec<TypeVar> {
+        self.0.read().expect("poisoned lexicon").free_vars()
     }
-}
-impl fmt::Display for TypeError {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match *self {
-            TypeError::Unification(ref e) => write!(f, "unification error: {}", e),
-            TypeError::OpNotFound => write!(f, "operator not found"),
-            TypeError::VarNotFound => write!(f, "variable not found"),
-        }
+    /// Sample a `Term`.
+    pub fn sample_term(
+        &mut self,
+        schema: &TypeSchema,
+        ctx: &mut Context,
+        invent: bool,
+        max_d: usize,
+    ) -> Result<Term, SampleError> {
+        let mut lex = self.0.write().expect("poisoned lexicon");
+        lex.sample_term(schema, ctx, invent, max_d, 0)
```

Comment:
This (and many following) could all be one-liners, right, or do they get too long?